File: grind75/Dec31/297.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Codec struct {
    
}

func Constructor() Codec {
    return Codec{}
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
    if root == nil { return "" }
    ret := []string{}
    q := []*TreeNode{root}

    for len(q) > 0 {
        lq := len(q)
        cq := []*TreeNode{}

        for lq > 0 {
            node := q[0]
            q = q[1:]
            lq--

            lv, rv := -9999, -9999
            if node.Left != nil { 
                lv = node.Left.Val
                cq = append(cq, node.Left)
            }
            if node.Right != nil { 
                rv = node.Right.Val
                cq = append(cq, node.Right)
            }
            ret = append(ret, fmt.Sprintf("%d/%d/%d", node.Val, lv, rv))
        }
        q = cq
    }
    // fmt.Println(strings.Join(ret, "@"))
    return strings.Join(ret, "@")
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {    
    if data == "" { return nil }
    decode := strings.Split(data, "@")
    root := &TreeNode{}
    q := []*TreeNode{root}

    for len(q) > 0 {
        cq := []*TreeNode{}
        lq := len(q)
        
        for lq > 0 {
            node := q[0]
            q = q[1:]
            info := decode[0]
            decode = decode[1:]
            lq--
            
            vals := strings.Split(info, "/")
            node.Val, _ = strconv.Atoi(vals[0])
            if vals[1] != "-9999" {
                val, _ := strconv.Atoi(vals[1])
                lnode := &TreeNode{Val: val}
                node.Left = lnode
                cq = append(cq, lnode)
            }
            if vals[2] != "-9999" {
                val, _ := strconv.Atoi(vals[2])
                rnode := &TreeNode{Val: val}
                node.Right = rnode
                cq = append(cq, rnode)
            }
        }
        q = cq
    }
    return root

}


/**
 * Your Codec object will be instantiated and called as such:
 * ser := Constructor();
 * deser := Constructor();
 * data := ser.serialize(root);
 * ans := deser.deserialize(data);
 */

File: grind75/Dec31/297.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:
    # using a preorder traversal
    # 
https://leetcode.com/problems/serialize-and-deserialize-binary-tree/solutions/394234/python-preorder/

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        def preorder(root):
            nonlocal traversal
            if not root:
                traversal += '#' + ' '
                return
            traversal += str(root.val) + ' '
            preorder(root.left)
            preorder(root.right)
        traversal = ''
        preorder(root)
        return traversal

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        data_list = data.split(' ')
        return self.helper(data_list)

    def helper(self, data_list):
        if data_list[0] == '#':
            data_list.pop(0)
            return

        root = TreeNode(data_list[0])
        data_list.pop(0)
        root.left = self.helper(data_list)
        root.right = self.helper(data_list)
        return root


        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))
